8327963: C2: fix construction of memory graph around Initialize node to prevent incorrect execution if allocation is removed

rwestrel opened the pull request:
An `Initialize` node for an `Allocate` node is created with a memory
`Proj` of adr type raw memory. In order for stores to be captured, the
memory state out of the allocation is a `MergeMem` with slices for the
various object fields/array element set to the raw memory `Proj` of
the `Initialize` node. If `Phi`s need to be created during later
transformations from this memory state, The `Phi` for a particular
slice gets its adr type from the type of the `Proj` which is raw
memory. If during macro expansion, the `Allocate` is found to have no
use and so can be removed, the `Proj` out of the `Initialize` is
replaced by the memory state on input to the `Allocate`. A `Phi` for
some slice for a field of an object will end up with the raw memory
state on input to the `Allocate` node. As a result, memory state at
the `Phi` is incorrect and incorrect execution can happen.

The fix I propose is, rather than have a single `Proj` for the memory
state out of the `Initialize` with adr type raw memory, to use one
`Proj` per slice added to the memory state after the `Initalize`. Each
of the `Proj` should return the right adr type for its slice. For that
I propose having a new type of `Proj`: `NarrowMemProj` that captures
the right adr type.

Logic for the construction of the `Allocate`/`Initialize` subgraph is
tweaked so the right adr type captured in is own `NarrowMemProj` is
added to the memory sugraph. Code that removes an allocation or moves
it also has to be changed so it correctly takes the multiple memory
projections out of the `Initialize` node into account.

One tricky issue is that when EA split types for a scalar replaceable
`Allocate` node:

1- the adr type captured in the `NarrowMemProj` becomes out of sync
  with the type of the slices for the allocation
  
2- before EA, the memory state for one particular field out of the
  `Initialize` node can be used for a `Store` to the just allocated
  object or some other. So we can have a chain of `Store`s, some to
  the newly allocated object, some to some other objects, all of them
  using the state of `NarrowMemProj` out of the `Initialize`. After
  split unique types, the `NarrowMemProj` is for the slice of a
  particular allocation. So `Store`s to some other objects shouldn't
  use that memory state but the memory state before the `Allocate`.
  
For that, I added logic to update the adr type of `NarrowMemProj`
during split unique types and update the memory input of `Store`s that
don't depend on the memory state out of the allocation site.

I also wrote a verification pass to check that, in the memory graph,
nodes on a particular slice have the right adr type. That verification
pass is not included here because it uncovered issues that are
unrelated to this particular issue so I intend to propose it (with
fixes for those other issues) separately.

I reused the test cases that Emanuel included in
https://github.com/openjdk/jdk/pull/18265 and went over all issues
linked to this one, tried the test cases and added the ones that I
could reproduce.


/cc hotspot-compiler

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warnings
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletColor16.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletColor32.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletMono16.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/beaninfo/images/JAppletMono32.gif)
&nbsp;⚠️ Patch contains a binary file (src/java.desktop/share/classes/javax/swing/doc-files/JRootPane-1.gif)
&nbsp;⚠️ Patch contains a binary file (test/jdk/javax/rmi/ssl/keystore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/javax/rmi/ssl/truststore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/jdk/internal/loader/URLClassPath/testclasses.jar)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/net/www/protocol/https/HttpsClient/dnsstore)
&nbsp;⚠️ Patch contains a binary file (test/jdk/sun/net/www/protocol/https/HttpsClient/ipstore)

### Issue
 * [JDK-8327963](https://bugs.openjdk.org/browse/JDK-8327963): C2: fix construction of memory graph around Initialize node to prevent incorrect execution if allocation is removed (**Bug** - P3)(⚠️ The fixVersion in this issue is [26] but the fixVersion in .jcheck/conf is 25, a new backport will be created when this pr is integrated.)


### Contributors
 * Emanuel Peter `<epeter@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24570/head:pull/24570` \
`$ git checkout pull/24570`

Update a local copy of the PR: \
`$ git checkout pull/24570` \
`$ git pull https://git.openjdk.org/jdk.git pull/24570/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24570`

View PR using the GUI difftool: \
`$ git pr show -t 24570`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24570.diff">https://git.openjdk.org/jdk/pull/24570.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24570#issuecomment-2792492656)
</details>

rwestrel commented:
/contributor add @eme64

merykitty commented:
It would be great if we have union memory slices for this.

rwestrel commented:
> It would be great if we have union memory slices for this.

Something like that would fix it but it would be trickier to get right that this point fix, I think. Do you see any other use for it?

turbanoff commented on test/hotspot/jtreg/compiler/macronodes/TestEliminationOfAllocationWithoutUse.java:
```diff
@@ -0,0 +1,338 @@
+        test1(arr1, arr2);
+
+        double sum = 0;
+        for(int i = 0; i < arr1.length; ++i) {
```
```suggestion
        for (int i = 0; i < arr1.length; ++i) {
```

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4120,6 +4120,22 @@ Node* ConnectionGraph::find_inst_mem(Node *orig_mem, int alias_idx, GrowableArra
+    if (!is_instance && result->Opcode() == Op_NarrowMemProj) {
+      // Memory for non known instance can safely skip over a known instance allocation (that memory state doesn't access
+      // the result of an allocation for a known instance).
+      assert(result->as_Proj()->_con == TypeFunc::Memory, "a NarrowMemProj can only be a memory projection");
```
Can we verify that already in the `NarrowMemProj`, i.e. its constructor?

rwestrel replied:
`TypeFunc::Memory` is now hardcoded in the `NarrowMemProj` constructor.

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4120,6 +4120,22 @@ Node* ConnectionGraph::find_inst_mem(Node *orig_mem, int alias_idx, GrowableArra
+      // the result of an allocation for a known instance).
+      assert(result->as_Proj()->_con == TypeFunc::Memory, "a NarrowMemProj can only be a memory projection");
+      assert(toop != nullptr, "");
+      Node *proj_in = result->in(0);
```
```suggestion
      Node* proj_in = result->in(0);
```

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4436,6 +4452,24 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+        InitializeNode* init = alloc->as_Allocate()->initialization();
+        assert(init != nullptr, "can't find Initialization node for this Allocate node");
+        for (DUIterator_Fast imax, i = init->fast_outs(imax); i < imax; i++) {
+          ProjNode* proj = init->fast_out(i)->as_Proj();
+          if (proj->Opcode() == Op_NarrowMemProj) {
```
```suggestion
          ProjNode* proj = init->fast_out(i)->as_NarrowMemProj();
          if (proj != nullptr) {
```
That way you don't need to do the hacky cast below `((NarrowMemProjNode*)proj)`.

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4436,6 +4452,24 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+            const TypePtr* adr_type = proj->adr_type();
+            const TypePtr* new_adr_type = tinst->add_offset(adr_type->offset());
+            if (adr_type != new_adr_type) {
+              uint alias_ix = _compile->get_alias_index(new_adr_type);
+              assert(_compile->get_general_index(alias_ix) == _compile->get_alias_index(adr_type), "new adr type should be narrowed down from existing adr type");
```
```suggestion
              DEBUG_ONLY( uint alias_idx = _compile->get_alias_index(new_adr_type); )
              assert(_compile->get_general_index(alias_idx) == _compile->get_alias_index(adr_type), "new adr type should be narrowed down from existing adr type");
```

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4436,6 +4452,24 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+            if (adr_type != new_adr_type) {
+              uint alias_ix = _compile->get_alias_index(new_adr_type);
+              assert(_compile->get_general_index(alias_ix) == _compile->get_alias_index(adr_type), "new adr type should be narrowed down from existing adr type");
+              igvn->hash_delete(proj);
+              ((NarrowMemProjNode*)proj)->set_adr_type(new_adr_type);
+              igvn->hash_insert(proj);
+              record_for_optimizer(proj);
```
That seems to be the only reason why we need the `_adr_type` non constant. How bad would it be if we just created a new node? Hmm maybe not worth it. Just wanted to know if you had considered it?

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4699,11 +4733,13 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+    } else if (n->is_MemBar()) { // MemBar nodes
+      if (!n->is_Initialize()) { // memory projections for Initialize pushed below (so we get to all their uses)
+        // we don't need to do anything, but the users must be pushed
+        n = n->as_MemBar()->proj_out_or_null(TypeFunc::Memory);
+        if (n == nullptr) {
+          continue;
+        }
```
Could we now have multiple `NarrowMemProj`? If so, what would happen here?

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4830,7 +4873,27 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
       }
+      if (mem->Opcode() == Op_NarrowMemProj) {
+        const TypePtr* at = mem->adr_type();
+        uint idx = (uint) _compile->get_alias_index(at->is_ptr());
```
```suggestion
        uint alias_idx = (uint) _compile->get_alias_index(at->is_ptr());
```
for consistency with the other code

eme64 replied:
Well, actually it is already being used inconsistently... hmm. More expressive may still be better. Up to you.

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4830,7 +4873,27 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+            cur = mem;
+          }
+        } else {
+          // projection for a known allocation on a non known allocation slice: skip over the allocation
```
The thing about "non known" sounds like we should not be able to skip it... But I guess we still know they are from unrelated slices somehow?

eme64 commented on src/hotspot/share/opto/graphKit.cpp:
```diff
@@ -3630,14 +3630,17 @@ Node* GraphKit::set_output_for_allocation(AllocateNode* alloc,
+    // Add an edge in the MergeMem for the header fields so an access to one of those has correct memory state
+    // Use one NarrowMemProjNode per slice to properly record the adr type of each slice. The Initialize node will have
+    // multiple projection as a result.
+    set_memory(_gvn.transform(new NarrowMemProjNode(init, TypeFunc::Memory, C->get_adr_type(mark_idx))), mark_idx);
+    int klass_idx = C->get_alias_index(oop_type->add_offset(oopDesc::klass_offset_in_bytes()));
+    set_memory(_gvn.transform(new NarrowMemProjNode(init, TypeFunc::Memory, C->get_adr_type(klass_idx))), klass_idx);
```
Hmm, so we now do have multiple projections for `TypeFunc::Memory`. That makes me a little nervous about `proj_out_or_null(TypeFunc::Memory)` elsewhere in your changes.

eme64 commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -1017,18 +1016,19 @@ void PhaseMacroExpand::process_users_of_allocation(CallNode *alloc) {
-          Node *mem = init->in(TypeFunc::Memory);
+        Node* mem = init->in(TypeFunc::Memory);
 #ifdef ASSERT
+        Node* mem_proj = init->proj_out_or_null(TypeFunc::Memory);
+        if (mem_proj != nullptr) {
```
What happens if there are multiple?

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -5450,6 +5451,24 @@ Node* InitializeNode::complete_stores(Node* rawctl, Node* rawmem, Node* rawptr,
   return rawmem;
 }
 
+void InitializeNode::replace_mem_projs_by(Node* mem, Compile* C) {
+  for (DUIterator_Fast imax, i = fast_outs(imax); i < imax; i++) {
+    ProjNode* proj = fast_out(i)->as_Proj();
+    if (proj->_con == TypeFunc::Memory) {
+      C->gvn_replace_by(proj, mem);
+    }
+  }
+}
+
+void InitializeNode::replace_mem_projs_by(Node* mem, PhaseIterGVN* igvn) {
+  for (DUIterator_Fast imax, i = fast_outs(imax); i < imax; i++) {
+    ProjNode* proj = fast_out(i)->as_Proj();
+    if (proj->_con == TypeFunc::Memory) {
+      igvn->replace_node(proj, mem);
+      --i; --imax;
+    }
+  }
+}
```
Can you please add a comment to the code why we need both variants?

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -105,4 +105,33 @@ class ProjNode : public Node {
+  virtual uint hash() const {
+    return ProjNode::hash() + _adr_type->hash();
+  }
+  virtual bool cmp(const Node &n) const {
```
```suggestion
  virtual bool cmp(const Node& n) const {
```

eme64 replied:
Subjective, up to you

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -105,4 +105,33 @@ class ProjNode : public Node {
+    _adr_type = adr_type;
+  }
+  virtual int Opcode() const;
+};
```
Ah, or does it already get printed from the `adr_type()` i.e. the virtual method?

eme64 replied:
Hmm, no I don't think so... right?

rwestrel replied:
Done in new commit.

eme64 commented on test/hotspot/jtreg/compiler/macronodes/TestEliminationOfAllocationWithoutUse.java:
```diff
@@ -0,0 +1,338 @@
+
+/*
+ * @test
+ * @bug 8327012
```
```suggestion
 * @bug 8327012 8327963
```
I would add the current bug id as well.

eme64 commented on test/hotspot/jtreg/compiler/macronodes/TestEliminationOfAllocationWithoutUse.java:
```diff
@@ -0,0 +1,338 @@
+ *          goes to the initial state, even if before the allocation there were stores on the array
+ *          slice. This leads to a messed up memory graph, and missing stores in the generated code.
+ *
+ * @run main/othervm -Xcomp -XX:-TieredCompilation
+ *                   -XX:CompileCommand=compileonly,compiler.macronodes.TestEliminationOfAllocationWithoutUse::test*
+ *                   compiler.macronodes.TestEliminationOfAllocationWithoutUse
+ * @run main/othervm -Xcomp
+ *                   -XX:CompileCommand=compileonly,compiler.macronodes.TestEliminationOfAllocationWithoutUse::test*
+ *                   compiler.macronodes.TestEliminationOfAllocationWithoutUse
```
Would a run without Xcomp make sense?

eme64 commented on test/hotspot/jtreg/compiler/macronodes/TestInitializingStoreCapturing.java:
```diff
@@ -0,0 +1,59 @@
+
+/*
+ * @test
+ * @bug 8327012
```
```suggestion
 * @bug 8327012 8327963
```

eme64 reviewed (changes requested):
@rwestrel  Thanks for looking into that. I vaguely remember working on https://github.com/openjdk/jdk/pull/18265 and being quite confused about the way we treat memory slices :sweat_smile: 

I think this seems like a good step in the right direction. Though I am a little unsure about the effect on `proj_out_or_null`, as you can see below. Maybe we can have a simple query that just checks `has_proj(TypeFunc::Memory)`? But there are also some cases where you actually use the projection, and I'm not sure if that means you might be missing some if there are multiple?

Like @merykitty I wonder if there are other cases where we have similar issues with memory slices. But I guess we should tackle those separately.

robcasloz commented:
Thanks for working on this, Roland! A "dumb" question: could the issue also be addressed by ensuring that dead allocations are removed earlier (e.g. in the call to `PhaseMacroExpand::eliminate_allocate_node` performed as part of escape analysis/scalar replacement, before loop optimizations)? It seems this would also prevent the miscompilations in `TestEliminationOfAllocationWithoutUse`, no?

rwestrel commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4120,6 +4120,22 @@ Node* ConnectionGraph::find_inst_mem(Node *orig_mem, int alias_idx, GrowableArra
       }
       result = result->in(MemNode::Memory);
     }
+    if (!is_instance && result->Opcode() == Op_NarrowMemProj) {
```
Done in new commit.

rwestrel commented:
I reworked this quite a bit.

1. In line with Emanuel's comment about making `_adr_type` `const` in `NarrowMemProj`, I changed the code in EA so, rather than update in place the `_adr_type` of the existing `NarrowMemProj`s, it creates new ones with the new `_adr_type`. That simplifies things because rather than replace existing `NarrowMemProj`s with new ones (which is what the update in place did in practice), the new code adds extra `NarrowMemProj`s and doesn't remove the existing ones. That, in turn, simplifies the logic elsewhere in EA because there's no need for the memory state that's not specific to a known allocations to be rewired around known allocations.

2. I added asserts to catch cases where `proj_out` is called but the node has more than one matching projection. With those asserts, I caught some false positive/cases where we got lucky and worked around them by reworking the code so it doesn't use `proj_out`. That's the case in `PhaseIdealLoop::intrinsify_fill()`: we can end up there with more than one `FramePtr` projection because the code pattern used elsewhere is to add one more projection and let identical projections common during igvn. Also in `PhaseIdealLoop::fix_ctrl_uses()`, the loop exit can have 3 projections in some cases. That's a transitory state before everything is wired correctly (a cases where we are lucky).

3. I tried to abstract away the many ways we iterate over projections in this patch.

rwestrel commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5542,8 +5542,19 @@ void LibraryCallKit::arraycopy_move_allocation_here(AllocateArrayNode* alloc, No
+      ProjNode* proj = init->fast_out(i)->as_Proj();
+      if (proj->_con == TypeFunc::Memory) {
+        int alias_idx = C->get_alias_index(proj->adr_type());
+        assert(alias_idx == Compile::AliasIdxRaw || alias_idx == elemidx || alias_idx == mark_idx || alias_idx == klass_idx, "should be raw memory or array element type");
```
`assert` is a nop if `ASSERT` not defined. Does that answer your question?

galderz replied:
Yup, thanks

rwestrel commented:
@eme64 ready for another round of reviews when you have time

robcasloz commented:
> > Thanks for working on this, Roland! A "dumb" question: could the issue also be addressed by ensuring that dead allocations are removed earlier (e.g. in the call to `PhaseMacroExpand::eliminate_allocate_node` performed as part of escape analysis/scalar replacement, before loop optimizations)? It seems this would also prevent the miscompilations in `TestEliminationOfAllocationWithoutUse`, no?
> 
> I don't thing that would work (but haven't tried). The problem is that the memory graph is broken around `Initialize` nodes from the time they are added to the IR (parse time) and that's only exposed once they are removed from the graph. But I don't think it matters when they are removed.

I gave it a quick try (https://github.com/openjdk/jdk/commit/c28f81a7ef2a4f3d3cb761ea23a80c09276e7e58) and removing dead array allocations early (which are trivially detected as non-escaping but marked as non-scalar replaceable) is sufficient to fix all failures in `TestEliminationOfAllocationWithoutUse`. I have not tested the patch thoroughly and it might be that the dead test is too weak or I am missing something else, but this seems to me something worth exploring before committing to the solution proposed in this PR.

> What you suggest does also sound quite conservative: it could very well be that an allocation looses all its uses after some rounds of optimization but in the scheme you suggest, that allocation wouldn't be optimized out.

Note that I am not necessarily suggesting disabling "late" elimination of allocations at macro expansion. But it would be good, in light of the above findings, to find actual cases where the seemingly simpler alternative of removing dead allocations early is not sufficient for correctness, to motivate the more complex approach proposed in this PR.

robcasloz commented:
> > Note that I am not necessarily suggesting disabling "late" elimination of allocations at macro expansion. But it would be good, in light of the above findings, to find actual cases where the seemingly simpler alternative of removing dead allocations early is not sufficient for correctness, to motivate the more complex approach proposed in this PR.
> 
> What happens with this bug is that a Phi created sometime after parsing inherits the type of the projection out of the Initialize which is wrong. No issue happens until the allocation is removed though. Only having allocations be removed early one shortens the window where bad things (a new Phi) can happen. But bad things could still happen. After all, we do some loop opts to help EA so maybe a similar issue could happen there. Or maybe, down the road, someone will change the way we do loop opts during EA because it helps EA and the bug will be back but we don't necessarily notice it until it happens at a user's site.
> 
> Beyond that, you're suggesting restricting elimination of allocation. What if, down the road, someone notices that it gets in the way of some other optimization? Then that someone we'll have to reconstruct the history here.
> 
> There's a history in c2 of fixing issues that are complicated by working around them. Often what happens is that we, later on, realize that the first work around wasn't sufficient and try to pile on more arounds. I also already ran into situations where everything to perform the needed optimization is there but it's disabled for some reason in a particular case and it's unclear why.
> 
> I'm in favor of fixing issues once and for all by targeting their root cause and trying to not sacrifice performance whenever possible. Truth is we have a limited view in what people are running though a set of microbenchmarks but we can't be sure something has no performance impact only because that set of microbenchmarks doesn't regress. Beyond that, what can appear as a safer workaround today could be more trouble down the line and in the end cause more confusion and work.

Thanks for the elaborate reply, Roland. I am also in favor of fundamental fixes instead of accumulating point fixes over time. Note that I do not suggest restricting elimination of dead array allocations. I guess I am mostly surprised to learn that we do not eliminate them as early as we can, but I understand your concern that similar issues may arise even if we did. I still think it would be good to include test cases to confirm that these are not only theoretical concerns, but that should not block the progress of this PR.

I also think it would be good to investigate, separately, early elimination of dead array allocations, even after the integration of this work. Dead allocations may inhibit later optimizations so it would be good to eliminate them as early as possible anyway. One difficulty (not addressed in https://github.com/openjdk/jdk/commit/c28f81a7ef2a4f3d3cb761ea23a80c09276e7e58) is that early array elimination should still generate the nonnegative array size check code.

rwestrel commented:
> I still think it would be good to include test cases to confirm that these are not only theoretical concerns, but that should not block the progress of this PR.

Here is a test case:

```
import java.util.Arrays;

public class TestAllocNoUseBadMemoryState {
    private static volatile int volatileField;

    public static void main(String[] args) {
        boolean[] allTrue = new boolean[3];
        Arrays.fill(allTrue, true);
        A a = new A();
        boolean[] allFalse = new boolean[3];
        for (int i = 0; i < 20_000; i++) {
            a.field1 = 0;
            test1(a, allTrue);
            test1(a, allFalse);
            if (a.field1 != 42) {
                throw new RuntimeException("Lost Store");
            }
        }
    }

    private static void test1(A otherA, boolean[] flags) {
        if (flags == null) {
        }
        otherA.field1 = 42;
        for (int i = 0; i < 3; i++) {
            A a = new A();
            if (flags[i]) {
                break;
            }
        }
    }

    private static class A {
        int field1;
    }
}
```
where all the damage is done early on when EA runs. A pass of loop opts before EA fully unrolls the loop and creates memory `Phi`s with incorrect `adr_type` (raw memory). Then EA removes the allocation. All that keeps the `Store` to `field1` alive then is uncommon traps from template predicates. Once they are removed, the `Store` goes away (first round of loop opts after EA).

I'll add that test case to the PR.

rwestrel commented:
> I also think it would be good to investigate, separately, early elimination of dead array allocations, even after the integration of this work. Dead allocations may inhibit later optimizations so it would be good to eliminate them as early as possible anyway. One difficulty (not addressed in [c28f81a](https://github.com/openjdk/jdk/commit/c28f81a7ef2a4f3d3cb761ea23a80c09276e7e58)) is that early array elimination should still generate the nonnegative array size check code.

That makes sense. It would be useful to have a bugs to track that one.

robcasloz commented:
> > I still think it would be good to include test cases to confirm that these are not only theoretical concerns, but that should not block the progress of this PR.
> 
> Here is a test case:
> 
> ```
> import java.util.Arrays;
> 
> public class TestAllocNoUseBadMemoryState {
>     private static volatile int volatileField;
> 
>     public static void main(String[] args) {
>         boolean[] allTrue = new boolean[3];
>         Arrays.fill(allTrue, true);
>         A a = new A();
>         boolean[] allFalse = new boolean[3];
>         for (int i = 0; i < 20_000; i++) {
>             a.field1 = 0;
>             test1(a, allTrue);
>             test1(a, allFalse);
>             if (a.field1 != 42) {
>                 throw new RuntimeException("Lost Store");
>             }
>         }
>     }
> 
>     private static void test1(A otherA, boolean[] flags) {
>         if (flags == null) {
>         }
>         otherA.field1 = 42;
>         for (int i = 0; i < 3; i++) {
>             A a = new A();
>             if (flags[i]) {
>                 break;
>             }
>         }
>     }
> 
>     private static class A {
>         int field1;
>     }
> }
> ```
> 
> where all the damage is done early on when EA runs. A pass of loop opts before EA fully unrolls the loop and creates memory `Phi`s with incorrect `adr_type` (raw memory). Then EA removes the allocation. All that keeps the `Store` to `field1` alive then is uncommon traps from template predicates. Once they are removed, the `Store` goes away (first round of loop opts after EA).
> 
> I'll add that test case to the PR.

Thanks Roland for taking the time to research this, this failure really illustrates why the general solution proposed by this PR is needed.

robcasloz commented on test/hotspot/jtreg/compiler/macronodes/TestEarlyEliminationOfAllocationWithoutUse.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private static void test1(A otherA, boolean[] flags) {
+        if (flags == null) {
+        }
```
Consider removing these two lines, which do not seem essential to reproduce the issue. Removing them also gets rid of the deoptimization, which simplifies the failure analysis.

rwestrel replied:
Good catch. Done in new commit.

robcasloz commented:
> > I also think it would be good to investigate, separately, early elimination of dead array allocations, even after the integration of this work. Dead allocations may inhibit later optimizations so it would be good to eliminate them as early as possible anyway. One difficulty (not addressed in [c28f81a](https://github.com/openjdk/jdk/commit/c28f81a7ef2a4f3d3cb761ea23a80c09276e7e58)) is that early array elimination should still generate the nonnegative array size check code.
> 
> That makes sense. It would be useful to have a bugs to track that one.

Turns out there is one already: [JDK-8180290](https://bugs.openjdk.org/browse/JDK-8180290), I just added a comment there.

eme64 commented:
> > > I also think it would be good to investigate, separately, early elimination of dead array allocations, even after the integration of this work. Dead allocations may inhibit later optimizations so it would be good to eliminate them as early as possible anyway. One difficulty (not addressed in [c28f81a](https://github.com/openjdk/jdk/commit/c28f81a7ef2a4f3d3cb761ea23a80c09276e7e58)) is that early array elimination should still generate the nonnegative array size check code.
> > 
> > 
> > That makes sense. It would be useful to have a bugs to track that one.
> 
> Turns out there is one already: [JDK-8180290](https://bugs.openjdk.org/browse/JDK-8180290), I just added a comment there.

Should we link it on JIRA?

robcasloz commented:
> > Turns out there is one already: [JDK-8180290](https://bugs.openjdk.org/browse/JDK-8180290), I just added a comment there.
> 
> Should we link it on JIRA?

Sure, done.

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -49,6 +49,89 @@ class MultiNode : public Node {
+    }
+  };
+
+  // Iterate with i over all Proj uses calling callback
+  template<class Callback, class Iterator> ProjNode* apply_to_projs_any_iterator(Iterator i, Callback callback) const {
+    for (; i.cont(); i.next()) {
+      Node* p = i.current();
+      if (p->is_Proj()) {
+        ProjNode* proj = p->as_Proj();
+        if (callback(proj)) {
+          return proj;
+        }
+      } else {
+        assert(p == this && is_Start(), "else must be proj");
+      }
+    }
+    return nullptr;
+  }
```
I was a little confused about this `apply_to_proj` construct. I this something we already use, a familiar concept, the `apply_to`?

Do we really iterate over `all` Proj uses? It seems that is not true if `callback` returns true, right?
This makes the semantic of the `apply_to_projs` a little subtle, right?

A suggestion: You could expect from `Callback` that it does not return `true/false`, but some `enum` value that describes what happens when you return that value. Maybe `ApplyToProjs::CONTINUE` vs `ApplyToProjs::BREAK_AND_RETURN_CURRENT_PROJ`?

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -105,4 +188,48 @@ class ProjNode : public Node {
+
+template <class Callback> ProjNode* MultiNode::apply_to_projs(DUIterator_Fast& imax, DUIterator_Fast& i, Callback callback, uint which_proj) const {
+  auto filter = [which_proj, callback](ProjNode* proj) {
+    if (proj->_con == which_proj && callback(proj)) {
+      return true;
+    }
+    return false;
```
Could also be simplified to a single line `return ...`

eme64 commented on test/hotspot/jtreg/compiler/macronodes/TestEliminationOfAllocationWithoutUse.java:
```diff
@@ -0,0 +1,338 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
Is the copyright year accurate?

rwestrel replied:
It's your test that took over and finished so you tell me: do you want the copyright updated?

eme64 replied:
Ah right. Well :see_no_evil:  I guess you could write `2024, 2025` then :)

eme64 commented on test/hotspot/jtreg/compiler/macronodes/TestInitializingStoreCapturing.java:
```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
```
Is the copyright year accurate?

robcasloz reviewed (changes requested):
In the common case where allocations are not eliminated, matching transforms the introduced `NarrowMemProj` nodes into a sequence of redundant, raw `MemProj` nodes, see e.g. B6 here: [after-gcm.pdf](https://github.com/user-attachments/files/20477560/after-gcm.pdf). Would it be possible to clean them up during matching (or perhaps already during, or right after, macro expansion)?

rwestrel commented on src/hotspot/share/opto/multnode.cpp:
```diff
@@ -45,30 +45,51 @@ Node *MultiNode::match( const ProjNode *proj, const Matcher *m ) { return proj->
-    }
-  }
-  return nullptr;
+  assert(number_of_projs(which_proj) <= 1, "only when there's a single projection");
```
So you're suggesting that this could return `nullptr` so the caller could then test for `nullptr` and have some fallback logic? I would stick with the assert: if C2 crashes because of this,  I think will be easier to diagnose an assert than an unexpected `nullptr` return.

rwestrel commented:
> I was a little confused about this apply_to_proj construct. I this something we already use, a familiar concept, the apply_to?

Thanks for the pointer to the style guide section. I followed the recommendation there. I also followed your suggestion for the return value of the callback.

All of your other comments should be addressed in new commit as well.

@eme64 please have another look when you find the time.

rwestrel commented on src/hotspot/share/opto/multnode.cpp:
```diff
@@ -45,30 +45,51 @@ Node *MultiNode::match( const ProjNode *proj, const Matcher *m ) { return proj->
-  }
-  return nullptr;
+  assert(number_of_projs(which_proj) <= 1, "only when there's a single projection");
+  auto find_proj = [which_proj, this](ProjNode* proj) {
```
Thanks for the report. This should be fixed now.

rwestrel commented:
> In the common case where allocations are not eliminated, matching transforms the introduced `NarrowMemProj` nodes into a sequence of redundant, raw `MemProj` nodes, see e.g. B6 here: [after-gcm.pdf](https://github.com/user-attachments/files/20477560/after-gcm.pdf). Would it be possible to clean them up during matching (or perhaps already during, or right after, macro expansion)?

Thanks for looking at this @robcasloz 
I made the change you requested.

robcasloz commented:
Since the changeset may introduce a significant amount of C2 nodes (especially for objects with many fields), I also evaluated the effect of this changeset on C2 speed (using DaCapo 23.11-chopin on x64 and aarch64). The result is neutral, I guess because 1) allocation C2 patterns are relatively infrequent and 2) the C2 back-end is unaffected (except for the rare pattern you discuss above where redundant memory projections are not removed, but even then register allocation is unaffected).

robcasloz commented:
> > > One way would be to simply assert that there's no `NarrowMemProj`s left during final graph reshape. Is that what you'd like?
> > 
> > 
> > Yes, that would be great (and I think it is OK to leave it to a future RFE if fully enforcing it would further increase the complexity of this PR).
> 
> Ok. I propose to file the follow up RFE if/once this PR has integrated as there's no follow up work until this code is in.

Sure, that makes sense, thanks.

rose00 commented:
> I think it would be good (although not necessarily in the context of this PR) to establish the "no duplicate memory projection" invariant in the back-end, for sanity and to make sure we do not break any logic that might be implicitly relying on it. If you agree, could you file a follow-up RFE, ideally with a reproducer where the current logic fails to remove `NarrowMemProj`s?

I see this as a request for a better "normal form" for the graph.  The trick here is that, if we are allowing temporary "abnormal" forms of the graph, in order to give various transforms some "working room" to rearrange things, we need to decide when are the moments when the graph must be settled back down into a normal form.

We sometimes check for some kinds of IR normality, and/or enforce some normality, in the "final graph reshape" phase.  The problem with loading up too many ad hoc operations at that point is, it may create a completely new kind of graph with new invariants.  (Don't like the current standard?  Create a new one, and see how that goes!  Same for global IR contracts.)  

Having two kinds of IR with two sets of invariants (one set more restrictive) has an obvious objection:  We fragment our ability to enforce the rules; we need to write enforcement logic which says "which phase are we in?" before checking the right set of rules.  And if the editing sessions are rare, we don't get much benefit from the rules that are enforced by that editing session.  By definition "final graph reshape" is rare.  It's worth it since we are going to a lower IR, which really must have different rules, but it's not a light thing to add to the design.

In any case, adding a normalization requirement seems to need a "wash pass" of some sort over the whole graph, to do necessary cleanups.  We do this sometimes, I think, after loop opts or EA, maybe other places, and at "final graph reshape".  This is going to be a runtime expense, I think, unless it can be piggybacked on some other pass we already do.  Maybe a hallmark of these "post-operative" cleanups is that the operation itself required some side data structure, created just for the operation (loop nest or connection graph) and discarded later in order to unleash unconstrained downstream transforms.  During the operation, transforms are specialized just to keep the side data structure relevant.  Afterwards, the graph "opens up" to unconstrained changes.  But in all cases, local updates should be as free as possible, even if their order varies randomly due to worklist artifacts, etc.  (BTW, this is why stress tests on worklist order are valuable.)

I'm not advocating firmly for or against new normalizations, but here's a final thought to throw in:  Performing normalizations seems to distrust an important design decision, noted in my previous comment.  That is, IR transformations should a "confluent" or "commutative"; it should not matter in what order you perform the transforms; you should still get to a better program, with identical user-visible semantics, whichever order you apply the transforms.  (Worklist stress tests, again…)  Obviously we globally schedule our tactics at the top level, but (in a deep sense) it should rarely matter what order we schedule things, at least as far as correctness goes.  And down in the details, at a local level, it really should not matter what form of graph you are working with.  Specifically, if we are using narrow memory projections sometimes, we should be prepared to respect them always. (Except, perhaps, at very well defined global cut-points, like final graph reshape, or a comprehensive cleanup after episodes of loop opts or EA.)

It's been a while since I coded C2 stuff as my day job, but HTH.

rwestrel commented:
/open

robcasloz commented:
> Specifically, if we are using narrow memory projections sometimes, we should be prepared to respect them always.

@rose00 I fully agree with your general argument, but note that my request refers to avoiding redundant MachProj memory projections arising after matching narrow memory projections (such as nodes 56-39 in B6 in the following CFG: https://github.com/user-attachments/files/20477560/after-gcm.pdf), not narrow memory projections per se. I do not see any use in allowing redundant MachProj memory projections in the IR, while due to their ambiguity they increase the risk of introducing new bugs or unveiling latent bugs, e.g. in anti-dependency analysis. So I am happy that Roland has found a cheap way to prevent them from ever appearing in the IR.

rwestrel commented:
@robcasloz thanks for the review. New commit addresses most of your comments.

rwestrel commented:
/contributor add @robcasloz

rwestrel commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -1604,7 +1603,7 @@ void PhaseMacroExpand::expand_initialize_membar(AllocateNode* alloc, InitializeN
       // elimination. Simply add the MemBarStoreStore after object
       // initialization.
-      MemBarNode* mb = MemBarNode::make(C, Op_MemBarStoreStore, Compile::AliasIdxBot);
+      MemBarNode* mb = MemBarNode::make(C, Op_MemBarStoreStore, Compile::AliasIdxRaw);
```
I updated the comment. Can you have a look?

robcasloz replied:
Thanks for the clarification and the pointer to the comment in `initialize_object`, that helps.

robcasloz commented:
Hi @rwestrel, could you please have a look at the merge conflicts of this PR so that we can progress further with the review of this work?

robcasloz commented:
> Hi @rwestrel, could you please have a look at the merge conflicts of this PR so that we can progress further with the review of this work?

The conflict is caused by the integration of [JDK-8360031](https://bugs.openjdk.org/browse/JDK-8360031), which relaxes the assertion in https://github.com/openjdk/jdk/blob/430041d366ddf450c2480c81608dde980dfa6d41/src/hotspot/share/opto/memnode.cpp#L4232 which is also touched by this changeset. Is the current assertion in mainline (after JDK-8360031) still valid in the context of this changeset?

robcasloz commented:
> Is the current assertion in mainline (after JDK-8360031) still valid in the context of this changeset?

I did a bit of testing and updating the asserted invariant to `(outcnt() > 0 && outcnt() <= 2) || Opcode() == Op_Initialize` seems to work.

eme64 commented on src/hotspot/share/opto/escape.cpp:
```diff
@@ -4441,6 +4446,25 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
+          }
+          return MultiNode::CONTINUE;
+        };
+        init->apply_to_narrow_mem_projs(i, process_narrow_proj);
```
Is there a reason why this is not a `init->for_each_narrow_mem_proj(callback)`, that has an internal iterator?

Because with this API, I'm wondering: What would happen if I feed `apply_to_narrow_mem_projs` an iterator that does not belong to the `init`?

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -5448,6 +5449,60 @@ Node* InitializeNode::complete_stores(Node* rawctl, Node* rawmem, Node* rawptr,
+    return BREAK_AND_RETURN_CURRENT_PROJ;
+  };
+  return apply_to_narrow_mem_projs(find_proj, adr_type) != nullptr;
+}
```
Am I seeing this right: `already_has_narrow_mem_proj_with_adr_type` calls `apply_to_narrow_mem_projs` with `callback = find_proj`. `find_proj` returns `BREAK_AND_RETURN_CURRENT_PROJ`, which is an element from `ApplyToProjs`. That would mean that when we call the `callback`, we get an enum element, and not a boolean, right?

If that is the case, you should probably not do an implicit comparison on this line:
`    if (proj->adr_type() == adr_type && callback(proj->as_NarrowMemProj())) {`
Hotspot style guide does not like implicit conversion. You should use an explicit comparison.

I think it would also be more clear what is happening. Currently, I'm a bit confused.

All the overloadings of `apply_to_narrow_mem_projs` make it a bit hard to see what goes where :/
I wonder if we really need all that complexity.

eme64 commented on src/hotspot/share/opto/memnode.hpp:
```diff
@@ -1388,6 +1401,31 @@ class InitializeNode: public MemBarNode {
+  // Run callback on all NarrowMem proj uses using passed iterator
+  template <class Callback> ProjNode* apply_to_narrow_mem_projs(DUIterator& i, Callback callback) const {
+    return apply_to_narrow_mem_projs_any_iterator<Callback, UsesIterator>(UsesIterator(i, this), callback);
+  }
```
The use in [ConnectionGraph::split_unique_types](https://github.com/openjdk/jdk/pull/24570/files#diff-03f7ae3cf79ff61be6e4f0590b7809a87825b073341fdbfcf36143b99c304474R4467) really does not need a return value, and also could work fine with a `callback` that return `void`.

eme64 commented on src/hotspot/share/opto/library_call.cpp:
```diff
@@ -5595,8 +5595,21 @@ void LibraryCallKit::arraycopy_move_allocation_here(AllocateArrayNode* alloc, No
+      set_memory(proj, alias_idx);
+      return MultiNode::CONTINUE;
+    };
+    init->apply_to_projs(move_proj, TypeFunc::Memory);
```
A "for each" using callback with `void` return would create a little less noise.

eme64 commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -1604,7 +1603,10 @@ void PhaseMacroExpand::expand_initialize_membar(AllocateNode* alloc, InitializeN
+      // See comment below in this if's other branch that explains why a raw memory MemBar is good enough. If init is
+      // null, this allocation does have an InitializeNode but this logic can't locate it (see comment in
+      // PhaseMacroExpand::initialize_object()).
+      MemBarNode* mb = MemBarNode::make(C, Op_MemBarStoreStore, Compile::AliasIdxRaw);
```
Would it not be nicer to have the explanation here than below, and refer from below to here? Would help the reading flow ;)

eme64 commented on src/hotspot/share/opto/macro.cpp:
```diff
@@ -1620,24 +1622,40 @@ void PhaseMacroExpand::expand_initialize_membar(AllocateNode* alloc, InitializeN
+        }
+        return MultiNode::CONTINUE;
+      };
+      init->apply_to_projs(find_raw_mem, TypeFunc::Memory);
```
Alternatives:
- `init->unique_raw_mem_proj()`
- `init->unique_out(raw_mem_proj_predicate)` with boolean predicate.
Both of those would assert if we find multiple or none.

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -5448,6 +5449,60 @@ Node* InitializeNode::complete_stores(Node* rawctl, Node* rawmem, Node* rawptr,
+    return CONTINUE;
+  };
+  apply_to_projs(imax, i, replace_proj, TypeFunc::Memory);
+}
```
Ouff, it's a little sad that we modify the iterator both outside and inside the method call. But I don't have a better solution now either. It may just be what we have to do.

eme64 commented on src/hotspot/share/opto/multnode.cpp:
```diff
@@ -45,30 +45,51 @@ Node *MultiNode::match( const ProjNode *proj, const Matcher *m ) { return proj->
+
+template<class Callback> ProjNode* MultiNode::apply_to_projs(Callback callback, uint which_proj, bool is_io_use) const {
+  auto filter = [&](ProjNode* proj) {
+    if (proj->_is_io_use == is_io_use && callback(proj)) {
```
implicit zero check?

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -106,6 +196,49 @@ class ProjNode : public Node {
+
+template <class Callback> ProjNode* MultiNode::apply_to_projs(DUIterator_Fast& imax, DUIterator_Fast& i, Callback callback, uint which_proj) const {
+  auto filter = [&](ProjNode* proj) {
+    if (proj->_con == which_proj && callback(proj)) {
```
Implicit zero check?

rwestrel commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -106,6 +196,49 @@ class ProjNode : public Node {
+#endif
+};
+
+template <class Callback> ProjNode* MultiNode::apply_to_projs(DUIterator_Fast& imax, DUIterator_Fast& i, Callback callback, uint which_proj) const {
```
It needs the `ProjNode` declaration because it accesses `proj->_con` and can't be in `multnode.cpp` because of the template parameter.

eme64 commented on src/hotspot/share/opto/memnode.cpp:
```diff
@@ -5445,6 +5449,60 @@ Node* InitializeNode::complete_stores(Node* rawctl, Node* rawmem, Node* rawptr,
+    return CONTINUE;
+  };
+  return apply_to_narrow_mem_projs(filter);
+}
```
It seems to me that the upper method is only used by the lower here. Why not just collapse them? It would also reduce the "overloading noise".

eme64 commented on src/hotspot/share/opto/memnode.hpp:
```diff
@@ -1388,6 +1401,46 @@ class InitializeNode: public MemBarNode {
+  // Run callback on all NarrowMem proj uses using passed iterator
+  template <class Callback> NarrowMemProjNode* apply_to_narrow_mem_projs(DUIterator& i, Callback callback) const {
+    return apply_to_narrow_mem_projs_any_iterator<Callback, UsesIterator>(UsesIterator(i, this), callback);
+  }
```
Basically we could call `apply_to_narrow_mem_projs_any_iterator` directly from the two uses:
- `already_has_narrow_mem_proj_with_adr_type`
- `for_each_narrow_mem_proj_with_new_uses`

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -49,6 +49,115 @@ class MultiNode : public Node {
+  }
+
+  // Same but for matching _con and _is_io_use
+  template <class Callback> ProjNode* apply_to_projs(Callback callback, uint which_proj, bool is_io_use) const;
```
It is just a lot of boilerplate, would be nice if it was a little slicker ;)

eme64 reviewed (commented):
@rwestrel Thanks for the updates, it already looks better :)

I had a few minutes to look over the `apply_..` solutions. I left a few comments, and hope that we can make the code just a little slicker still ;)

eme64 commented on src/hotspot/share/opto/memnode.hpp:
```diff
@@ -1388,6 +1401,32 @@ class InitializeNode: public MemBarNode {
+
+public:
+
+  template <class Callback> void for_each_narrow_mem_proj_with_new_uses(Callback callback) const {
```
Can you add a code comment what the "with new uses" part means? Probably that if we add more uses during iteration, we will eventually also visit those?

eme64 commented on src/hotspot/share/opto/multnode.hpp:
```diff
@@ -106,6 +196,49 @@ class ProjNode : public Node {
+    return sizeof(*this);
+  }
+public:
+  NarrowMemProjNode(Node* src, const TypePtr* adr_type)
```
@rwestrel Do you not like this suggestion?

eme64 commented on src/hotspot/share/opto/memnode.hpp:
```diff
@@ -1388,6 +1401,32 @@ class InitializeNode: public MemBarNode {
+      return nullptr;
+    }
+    return res->as_NarrowMemProj();
+  }
```
You are right. Ignore. My post-jogging brain saw things that were not there :laughing:
